src/hotspot/share/opto/loopnode.cpp

```diff
@@ -1616,6 +1616,88 @@ bool PhaseIdealLoop::convert_to_long_loop(Node* cmp, Node* phi, IdealLoopTree* l
 
 //------------------------------is_counted_loop--------------------------------
 bool PhaseIdealLoop::is_counted_loop(Node* x, IdealLoopTree*&loop, BasicType iv_bt) {
+  return do_is_counted_loop(x, loop, iv_bt) || is_counted_loop_with_speculative_long_limit(x, loop, iv_bt);
+}
+
+// When comparing an int incrementor with long limit, check if the limit is within int range. If so, treat it as an int
+// counted loop but deoptimize if the limit is out of int range.
+// This is common pattern with "for (int i =...; i < long_limit; ...)" where int "i" is implicitly promoted to long,
+// i.e., "(long) i < long_limit", and therefore making it not an int counted loop without this transformation.
+bool PhaseIdealLoop::is_counted_loop_with_speculative_long_limit(Node* x, IdealLoopTree*&loop, BasicType iv_bt) {
+  if (iv_bt != T_INT) {
+    return false;
+  }
+
+  Node* back_control = loop_exit_control(x, loop);
+  if (back_control == nullptr) {
+    return false;
+  }
+
+  Node* init_control = x->in(LoopNode::EntryControl);
+
+  // Make sure there is a parse predicate for us to insert the loop limit check.
+  const Predicates predicates(init_control);
+  const PredicateBlock* loop_limit_check_predicate_block = predicates.loop_limit_check_predicate_block();
+  if (!loop_limit_check_predicate_block->has_parse_predicate()) {
+    return false;
+  }
+
+  BoolTest::mask bt = BoolTest::illegal;
+  float cl_prob = 0;
+  Node* incr = nullptr;
+  Node* limit = nullptr;
+  Node* cmp = loop_exit_test(back_control, loop, incr, limit, bt, cl_prob);
+
+  // Check exit test is a "(long) i < long_limit" pattern.
+  if (cmp == nullptr || cmp->Opcode() != Op_CmpL || incr == nullptr || incr->Opcode() != Op_ConvI2L) {
+    return false;
+  }
+
+  ParsePredicateNode* loop_limit_check_parse_predicate = loop_limit_check_predicate_block->parse_predicate();
+  if (!is_dominator(get_ctrl(limit), loop_limit_check_parse_predicate->in(0))) {
+    return false;
+  }
+
+  // Replace exit test nodes. Need to revert changes if this still doesn't make it a counted loop.
+  Node* new_incr = incr->in(1);
+
+  // Optimistically transform "(long) i < long_limit" to "i < (int) long_limit".
+  Node* new_limit = _igvn.register_new_node_with_optimizer(new ConvL2INode(limit), limit);
+  set_early_ctrl(new_limit, get_ctrl(limit));
+
+  Node* new_cmp = _igvn.register_new_node_with_optimizer(
+      cmp->in(1) == incr
+          ? new CmpINode(new_incr, new_limit) : new CmpINode(new_limit, new_incr),
+      cmp);
```

Comment:
This is a little hard to read. How about:
```suggestion
  Node* new_cmp = cmp->in(1) == incr ? new CmpINode(new_incr, new_limit) : new CmpINode(new_limit, new_incr);
  _igvn.register_new_node_with_optimizer(new_cmp, cmp);
```